collie/interactions/datasets.py

```diff
@@ -333,45 +342,70 @@ def __init__(self,
             )
 
         # When an ``Interactions`` is instantiated with exact negative sampling, a
-        # ``positive_items`` attribute is created, a ``set`` of the ``mat`` object that enables
-        # fast, O(1), ``(row, col)`` lookup. When ``__getitem__`` is called, negative item IDs are
-        # sampled one-at-a-time from all possible values in ``self.num_items``, we check if that
-        # user ID, item ID pair is in ``self.positive_items``, and sample continuously until we
-        # have a negative match or reach a limit of ``max_number_of_samples_to_consider`` tries
+        # ``positive_sets`` attribute is created, a ``set`` of the ``mat`` object that enables
+        # fast, O(1), ``(row, col)`` lookup. When ``__getitem__`` is called, negative item or
+        # user IDs are sampled one-at-a-time from all possible values in ``self.num_items``
+        # or ``self.num_users``, we check if that user ID, item ID pair is in
+        # ``self.positive_sets``, and sample continuously until we have a negative match or
+        # reach a limit of ``max_number_of_samples_to_consider`` tries
         if self.check_num_negative_samples_is_valid:
             print('Checking ``num_negative_samples`` is valid...')
+            if self.negative_sample_type == 'item':
+                if self._checked_num_negative_item_samples_is_valid is False:
+                    self._checked_num_negative_item_samples_is_valid = (
+                        self._check_num_negative_samples_is_valid()
+                    )
+            elif self.negative_sample_type == 'user':
+                if self._checked_num_negative_user_samples_is_valid is False:
+                    self._checked_num_negative_user_samples_is_valid = (
+                        self._check_num_negative_samples_is_valid()
+                    )
+
+        self.positive_sets = {}
+        if self.max_number_of_samples_to_consider > 0:
+            print(f'Generating positive {self.negative_sample_type}s set...')
+            self._generate_positive_set()
+
+    def _check_num_negative_samples_is_valid(self) -> bool:
```

Comment:
🚨 Incoming length explanation 🚨 

In the future when we will have this be able to alternate between user and item negative sampling, we'll want to check that the sampling type is valid before officially changing it. If we don't, then someone can change it, validation fails, but it was still changed. 

With the current way this is setup, that logic would probably look like: 

```python
def _change_num_negative_samples_type(self, new_type: Literal['user', 'item']):
    self.negative_sample_type = new_type  # needs to be changed before we validate it
    self._check_num_negative_samples_is_valid()
```

I'm thinking this method can take in the actual negative sampling type so we can avoid this. 

```suggestion
    def _check_num_negative_samples_is_valid(self, negative_sample_type: Literal['user', 'item']) -> bool:
```

Does that make sense?

Reply:
Ah great point! I added your suggestion.